### PR TITLE
add amazon linux as a dist. uses docker instead of docker-ce as rpm dep

### PR DESCRIPTION
**Pull request by ramarnat**
Install [fails](https://github.com/NVIDIA/nvidia-docker/issues/618) on Amazon Linux AMIs.

### CODE REVIEW

**Comment by musedivision** [comment]
I would also like this added. Trying to get nvidia-docker setup on AWS p2.xlarge instance.

**Comment by flx42** [comment]
Do we really need to provide all versions of docker here?
Would 17.12 be enough?

**Comment by ramarnat** [comment]
I was following the convention that was established, and it was easy enough to replicate the block for the old docker versions. I don't have an issue with truncating that.

**Comment by ramarnat** [comment]
Question. Thoughts on the best way to integrate the updated libcap into the Amazon Linux rpm build? The make and compile could be built into the Dockerfile, but the spec would need to be modified to pick up the newer libcap?

```git clone https://kernel.googlesource.com/pub/scm/linux/kernel/git/morgan/libcap.git
cd libcap
make
sudo cp libcap/libcap.so.2.25 /lib64
sudo rm /lib64/libcap.so.2
sudo ln -s /lib64/libcap.so.2.25 /lib64/libcap.so.2```

**Comment by flx42** [comment]
@3XX0 will try to fix the library.

Looks like some of these old versions are not available, let's not build packages for versions that do not exist:
```
[ec2-user@ip-172-31-24-128 ~]$ yum search --showduplicates docker
Loaded plugins: priorities, update-motd, upgrade-helper
================================================================================= N/S matched: docker ==================================================================================
docker-storage-setup-0.5-1.7.gite193b3b.amzn1.noarch : A simple service to setup docker storage devices
docker-storage-setup-0.6.0-1.18.giteb688d4.amzn1.noarch : A simple service to setup docker storage devices
docker-17.03.2ce-1.59.amzn1.x86_64 : Automates deployment of containerized applications
docker-17.06.2ce-1.93.amzn1.x86_64 : Automates deployment of containerized applications
docker-17.06.2ce-1.94.amzn1.x86_64 : Automates deployment of containerized applications
docker-17.09.1ce-1.111.amzn1.x86_64 : Automates deployment of containerized applications
```
We can keep `17.12` though.

**Comment by flx42** [comment]
Also, maybe we should suffix the generated package versions with `amzn1` like they do. To make it more obvious when someone installs the wrong repo.

**Comment by flx42** [comment]
Please sign-off commit.

**Comment by spkaplan** [comment]
Sorry, I am not clear on the outcome. Is there a fix for this? I am having the same issue as https://github.com/NVIDIA/nvidia-docker/issues/618, with docker 17.09.1.

**Comment by ramarnat** [comment]
Sorry, I have been busy the past couple of weeks. Do you need me to do anything on my end?

**Comment by flx42** [comment]
Please sign-off the commit. :)

**Comment by ramarnat** [comment]
@flx42 you have my sign off - did not know if you needed me to update the commit to include some of your comments, which might be harder to do till about middle of march for me. 

my branch works well for the init script I wrote/adapted from a few sources for the ecs instances that run our jobs:
```
#!/usr/bin/env bash
set -x
# http://docs.aws.amazon.com/batch/latest/userguide/batch-gpu-ami.html
# http://docs.aws.amazon.com/batch/latest/userguide/example-job-definitions.html#example-test-gpu
# Install ecs-init, start docker, and install nvidia-docker
sudo yum install -y ecs-init
sudo service docker start

curl -s -L https://nvidia.github.io/nvidia-docker/centos7/x86_64/nvidia-docker.repo | \
>   sudo tee /etc/yum.repos.d/nvidia-docker.repo

git clone https://kernel.googlesource.com/pub/scm/linux/kernel/git/morgan/libcap.git
cd libcap
make
sudo cp libcap/libcap.so.2.25 /lib64
sudo rm /lib64/libcap.so.2
sudo ln -s /lib64/libcap.so.2.25 /lib64/libcap.so.2

git clone https://github.com:full360/nvidia-docker.git
cd nvidia-docker
make 17.09.1.ce-amzn1
cd dist/amzn1
sudo rpm -Uvh ./nvidia-docker2-2.0.2-1.docker17.09.1.ce.noarch.rpm

# Validate installation
sudo yum install -y nvidia-container-runtime-1.1.1-1.docker17.09.1.x86_64

# Make sure the NVIDIA kernel modules and driver files are bootstraped
# Otherwise running a GPU job inside a container will fail with "cuda: unknown exception"
echo '#!/bin/bash' | sudo tee /var/lib/cloud/scripts/per-boot/00_nvidia-modprobe > /dev/null
echo 'nvidia-modprobe -u -c=0' | sudo tee --append /var/lib/cloud/scripts/per-boot/00_nvidia-modprobe > /dev/null
sudo chmod +x /var/lib/cloud/scripts/per-boot/00_nvidia-modprobe
sudo /var/lib/cloud/scripts/per-boot/00_nvidia-modprobe
```

**Comment by spkaplan** [comment]
@flx42 From your point-of-view, are the changes ready to merge, or do you want @ramarnat to incorporate feedback from some of your comments (which he won't have time to do until mid-march)?

**Comment by flx42** [comment]
yes, it's ready. By sign-off I mean this:
https://git-scm.com/docs/git-commit

> -s
> --signoff
> 
>     Add Signed-off-by line by the committer at the end of the commit log message. The meaning of a signoff depends on the project, but it typically certifies that committer has the rights to submit this work under the same license and agrees to a Developer Certificate of Origin (see http://developercertificate.org/ for more information).

So, `git commit --amend --signoff` then force push to your branch.

**Comment by ramarnat** [comment]
ah, gotcha. Sorry hadnt done one of those before. Will take care of it and re-submit. I will strip the unwanted versions and see about adding the suffix.

**Comment by ramarnat** [comment]
@flx42 this one is signed and has the set of docker versions you listed. 

If you want to point me at where the suffix needs to be added, I can do that when I find a few minutes over the next couple of weeks.

**Comment by spkaplan** [comment]
@flx42, can you please explain in more detail where you would like the suffixes to unblock @ramarnat?

**Comment by flx42** [comment]
It's good now. I just need one final approval internally.
I will do the last change on top.

**Comment by spkaplan** [comment]
@flx42 Sorry for pestering. But do you have an estimate for when the final internal approval will happen? Thank you.

**Comment by flx42** [comment]
Should be this week.

**Comment by flx42** [comment]
I haven't forgotten, I hope I will be able to do it tomorrow.

File: Makefile [hunk 1]
```diff
@@ -170,3 +172,35 @@ stretch: 17.12.0-stretch 17.09.1-stretch 17.09.0-stretch 17.06.2-stretch 17.03.2
                         --build-arg PKG_REV="$(PKG_REV)" \
                         -t nvidia-docker2:$@ -f Dockerfile.stretch .
 	$(DOCKER) run --rm -v $(DIST_DIR)/stretch:/dist:Z nvidia-docker2:$@
+
+17.12.0.ce-amzn1:
```
Comment (fearphage) [inline Makefile]: FYI you could combine these into one rule.

```makefile
17.12.0.ce-amzn1 17.09.1.ce-amzn1 17.06.2.ce-amzn1 17.03.2.ce-amzn1:
	$(DOCKER) build --build-arg RUNTIME_VERSION="$(RUNTIME_VERSION)-1.docker$(subst .ce-amzn1,,$@)" \
                        --build-arg DOCKER_VERSION="docker = $(subst -amzn1,,$@)" \
                        --build-arg PKG_VERS="$(VERSION)" \
                        --build-arg PKG_REV="$(PKG_REV).docker$(subst -amzn1,,$@)" \
                        -t nvidia-docker2:$@ -f Dockerfile.amzn1 .
	$(DOCKER) run --rm -v $(DIST_DIR)/amzn1:/dist:Z nvidia-docker2:$@
```

File: Makefile [hunk 2]
```diff
@@ -11,12 +11,14 @@ DIST_DIR  := $(CURDIR)/dist
...
 .PHONY: all
 
-all: xenial centos7 stretch
+all: xenial centos7 stretch amzn1
 
 xenial: 17.12.0-xenial 17.09.1-xenial 17.09.0-xenial 17.06.2-xenial 17.03.2-xenial 1.13.1-xenial 1.12.6-xenial
 
 centos7: 17.12.0.ce-centos7 17.09.1.ce-centos7 17.09.0.ce-centos7 17.06.2.ce-centos7 17.03.2.ce-centos7 1.13.1-centos7 1.12.6-centos7
 
+amzn1: 17.12.0.ce-amzn1 17.09.1.ce-amzn1 17.09.0.ce-amzn1 17.06.2.ce-amzn1 17.03.2.ce-amzn1 1.13.1-amzn1 1.12.6-amzn1
```
Comment (fearphage) [inline Makefile]: `1.13.1-amzn1 1.12.6-amzn1` don't have targets unless I'm missing something.

File: Makefile (hunk 1 above)
Comment (flx42) [reply Makefile]: Not a big fan of this, but we could do that since it's getting a bit out of control.

Comment (fearphage) [reply Makefile]: It was merely a suggestion, but I'm just curious of what _specifically_ do you find unappealing?

**Comment by flx42** [comment]
@ramarnat sorry, small complication internally. nvidia-docker is still under [CLA](https://github.com/NVIDIA/nvidia-docker#issues-and-contributing) for now.
Do you want to sign the CLA? Or are you OK if I author the commits under my name? (+ a few fixes that I need to do).

**Comment by ramarnat** [comment]
feel free to author under your name. All good

**Comment by flx42** [comment]
Implemented, but pushing the packages will have to wait next week, sorry.

**Comment by flx42** [comment]
Packages are online, please test. Same as centos 7 instructions but with `amzn1` in the URL.

**Pull request closed without merging**